### PR TITLE
Type annotations for sqlalchemy.sql.selectable

### PR DESCRIPTION
**Pull request by dzmitar**
### Description
An attempt to annotate lib/sqlalchemy/sql/selectable.py with type hints (issue #6810)

### Checklist

This pull request is:

- [ ] A documentation / typographical error fix
	- Good to go, no issue or tests are needed
- [ ] A short code fix
	- please include the issue number, and create an issue if none exists, which
	  must include a complete example of the issue.  one line code fixes without an
	  issue and demonstration will not be accepted.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.   one line code fixes without tests will not be accepted.
- [ ] A new feature implementation
	- please include the issue number, and create an issue if none exists, which must
	  include a complete example of how the feature would look.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.

### CODE REVIEW

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision 59c8bbf866f1d2271bf2e1b9e5e22aa253024641 of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
New Gerrit review created for change 59c8bbf866f1d2271bf2e1b9e5e22aa253024641: https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4325

**Comment by zzzeek** [comment]
great!    We can probably fix a lot of those ignores but let's start with this

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision c156cae4551c3fa447cb5f27b424a7d85a1f45b2 of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
Patchset c156cae4551c3fa447cb5f27b424a7d85a1f45b2 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4325

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision 6c5c5bfd4138f22fd26b12e5ccd3c9c08e1b7890 of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
Patchset 6c5c5bfd4138f22fd26b12e5ccd3c9c08e1b7890 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4325

File: lib/sqlalchemy/sql/selectable.py [hunk 1]
```diff
@@ -1995,7 +2020,7 @@ class CTE(
     __visit_name__ = "cte"
 
     _traverse_internals: _TraverseInternalsType = (
```
Comment (sqla-tester) [inline lib/sqlalchemy/sql/selectable.py:2022]: **mike bayer** (zzzeek) wrote:

remove

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4325

File: lib/sqlalchemy/sql/selectable.py [hunk 2]
```diff
@@ -2067,7 +2092,7 @@ def alias(self, name: Optional[str] = None, flat: bool = False) -> CTE:
             :func:`_expression.alias`
 
         """
```
Comment (sqla-tester) [inline lib/sqlalchemy/sql/selectable.py:2094]: **mike bayer** (zzzeek) wrote:

remove

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4325

File: lib/sqlalchemy/sql/selectable.py [hunk 3]
```diff
@@ -2126,7 +2151,7 @@ def union_all(self, *other: _SelectStatementForCompoundArgument) -> CTE:
             self.element
         ), f"CTE element f{self.element} does not support union_all()"
```
Comment (sqla-tester) [inline lib/sqlalchemy/sql/selectable.py:2153]: **mike bayer** (zzzeek) wrote:

remove

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4325

File: lib/sqlalchemy/sql/selectable.py [hunk 4]
```diff
@@ -2755,7 +2780,7 @@ def cte(
             :meth:`_expression.HasCTE.cte`.
 
         """
```
Comment (sqla-tester) [inline lib/sqlalchemy/sql/selectable.py:2782]: **mike bayer** (zzzeek) wrote:

remove

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4325

File: lib/sqlalchemy/sql/selectable.py [hunk 5]
```diff
@@ -2860,7 +2887,7 @@ def alias(
     ) -> NamedFromGrouping:
         return NamedFromGrouping(self.element.alias(name=name, flat=flat))
```
Comment (sqla-tester) [inline lib/sqlalchemy/sql/selectable.py:2889]: **mike bayer** (zzzeek) wrote:

remove

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4325

File: lib/sqlalchemy/sql/selectable.py [hunk 6]
```diff
@@ -3427,11 +3454,11 @@ def exported_columns(
         "from, use the :attr:`_expression.SelectBase.selected_columns` "
         "attribute.",
     )
-    def c(self):
+    def c(self) -> ReadOnlyColumnCollection[str, KeyedColumnElement[Any]]:
         return self._implicit_subquery.columns
 
     @property
```
Comment (sqla-tester) [inline lib/sqlalchemy/sql/selectable.py:3460]: **mike bayer** (zzzeek) wrote:

remove

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4325

File: lib/sqlalchemy/sql/selectable.py [hunk 7]
```diff
@@ -3527,7 +3554,7 @@ def scalar_subquery(self) -> ScalarSelect[Any]:
         if self._label_style is not LABEL_STYLE_NONE:
             self = self.set_label_style(LABEL_STYLE_NONE)
```
Comment (sqla-tester) [inline lib/sqlalchemy/sql/selectable.py:3556]: **mike bayer** (zzzeek) wrote:

emove

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4325

File: lib/sqlalchemy/sql/selectable.py [hunk 8]
```diff
@@ -3553,7 +3580,7 @@ def lateral(self, name: Optional[str] = None) -> LateralFromClause:
             :ref:`tutorial_lateral_correlation` -  overview of usage.
 
         """
```
Comment (sqla-tester) [inline lib/sqlalchemy/sql/selectable.py:3582]: **mike bayer** (zzzeek) wrote:

remove

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4325

File: lib/sqlalchemy/sql/selectable.py [hunk 9]
```diff
@@ -3595,9 +3622,9 @@ def subquery(self, name: Optional[str] = None) -> Subquery:
...
```
Comment (sqla-tester) [inline lib/sqlalchemy/sql/selectable.py:3624]: **mike bayer** (zzzeek) wrote:

remove

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4325

File: lib/sqlalchemy/sql/selectable.py [hunk 10]
```diff
@@ -3643,11 +3670,11 @@ class SelectStatementGrouping(GroupedElement, SelectBase):
 
     element: SelectBase
 
-    def __init__(self, element):
+    def __init__(self, element: Select[Any]) -> None:
         self.element = coercions.expect(roles.SelectStatementRole, element)
 
-    def _ensure_disambiguated_names(self):
-        new_element = self.element._ensure_disambiguated_names()
+    def _ensure_disambiguated_names(self) -> SelectStatementGrouping:
```
Comment (sqla-tester) [inline lib/sqlalchemy/sql/selectable.py:3676]: **mike bayer** (zzzeek) wrote:

remove

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4325

File: lib/sqlalchemy/sql/selectable.py [hunk 11]
```diff
@@ -3660,11 +3687,12 @@ def set_label_style(
         self, label_style: SelectLabelStyle
     ) -> SelectStatementGrouping:
         return SelectStatementGrouping(
```
Comment (sqla-tester) [inline lib/sqlalchemy/sql/selectable.py:3689]: **mike bayer** (zzzeek) wrote:

remove

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4325

File: lib/sqlalchemy/sql/selectable.py [hunk 12]
```diff
@@ -4233,7 +4261,7 @@ def group_by(
 @CompileState.plugin_for("default", "compound_select")
 class CompoundSelectState(CompileState):
     @util.memoized_property
```
Comment (sqla-tester) [inline lib/sqlalchemy/sql/selectable.py:4263]: **mike bayer** (zzzeek) wrote:

remove

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4325

File: lib/sqlalchemy/sql/selectable.py [hunk 13]
```diff
@@ -4347,24 +4375,24 @@ def _scalar_type(self) -> TypeEngine[Any]:
     def self_group(
         self, against: Optional[OperatorType] = None
     ) -> GroupedElement:
```
Comment (sqla-tester) [inline lib/sqlalchemy/sql/selectable.py:4377]: **mike bayer** (zzzeek) wrote:

remove

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4325

File: lib/sqlalchemy/sql/selectable.py [hunk 14]
```diff
@@ -4347,24 +4375,24 @@ def _scalar_type(self) -> TypeEngine[Any]:
...
         if self._label_style is not style:
             self = self._generate()
             select_0 = self.selects[0].set_label_style(style)
             self.selects = [select_0] + self.selects[1:]
 
         return self
 
-    def _ensure_disambiguated_names(self):
-        new_select = self.selects[0]._ensure_disambiguated_names()
+    def _ensure_disambiguated_names(self) -> CompoundSelect:
```
Comment (sqla-tester) [inline lib/sqlalchemy/sql/selectable.py:4394]: **mike bayer** (zzzeek) wrote:

remove

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4325

File: lib/sqlalchemy/sql/selectable.py [hunk 15]
```diff
@@ -4548,31 +4576,34 @@ def _column_naming_convention(
...
+        ) -> Optional[str]:
             if c._is_text_clause:
                 return None
 
             elif not dedupe:
-                name = c._proxy_key
+                name = c._proxy_key  # type: ignore
                 if name is None:
                     name = "_no_label"
                 return name
```
Comment (sqla-tester) [inline lib/sqlalchemy/sql/selectable.py:4591]: **mike bayer** (zzzeek) wrote:

remove

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4325

File: lib/sqlalchemy/sql/selectable.py [hunk 16]
```diff
@@ -4548,31 +4576,34 @@ def _column_naming_convention(
...
 
-            name = c._tq_key_label if table_qualified else c._proxy_key
+            name = c._tq_key_label if table_qualified else c._proxy_key  # type: ignore # noqa: E501
 
             if name is None:
                 name = "_no_label"
                 if name in names:
```
Comment (sqla-tester) [inline lib/sqlalchemy/sql/selectable.py:4596]: **mike bayer** (zzzeek) wrote:

remove

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4325

File: lib/sqlalchemy/sql/selectable.py [hunk 17]
```diff
@@ -4548,31 +4576,34 @@ def _column_naming_convention(
...
-                    return c._anon_label(name) % pa
+                    return c._anon_label(name) % pa  # type: ignore
                 else:
                     names.add(name)
                     return name
 
             elif name in names:
                 return (
```
Comment (sqla-tester) [inline lib/sqlalchemy/sql/selectable.py:4603]: **mike bayer** (zzzeek) wrote:

remove

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4325

File: lib/sqlalchemy/sql/selectable.py [hunk 18]
```diff
@@ -4548,31 +4576,34 @@ def _column_naming_convention(
...
-                    c._anon_tq_key_label % pa
+                    c._anon_tq_key_label % pa  # type: ignore
                     if table_qualified
```
Comment (sqla-tester) [inline lib/sqlalchemy/sql/selectable.py:4605]: **mike bayer** (zzzeek) wrote:

remove

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4325

File: lib/sqlalchemy/sql/selectable.py [hunk 19]
```diff
@@ -4636,7 +4667,7 @@ def _normalize_froms(
         if froms:
             toremove = set(
                 itertools.chain.from_iterable(
```
Comment (sqla-tester) [inline lib/sqlalchemy/sql/selectable.py:4669]: **mike bayer** (zzzeek) wrote:

remove

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4325

File: lib/sqlalchemy/sql/selectable.py [hunk 20]
```diff
@@ -4684,8 +4715,8 @@ def _get_display_froms(
                     f
                     for f in froms
                     if f
```
Comment (sqla-tester) [inline lib/sqlalchemy/sql/selectable.py:4717]: **mike bayer** (zzzeek) wrote:

remove both

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4325

File: lib/sqlalchemy/sql/selectable.py [hunk 21]
```diff
@@ -4698,8 +4729,8 @@ def _get_display_froms(
                 f
                 for f in froms
                 if f
```
Comment (sqla-tester) [inline lib/sqlalchemy/sql/selectable.py:4731]: **mike bayer** (zzzeek) wrote:

remove both

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4325

File: lib/sqlalchemy/sql/selectable.py [hunk 22]
```diff
@@ -4716,7 +4747,7 @@ def _get_display_froms(
...
```
Comment (sqla-tester) [inline lib/sqlalchemy/sql/selectable.py:4749]: **mike bayer** (zzzeek) wrote:

remove

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4325

File: lib/sqlalchemy/sql/selectable.py [hunk 23]
```diff
@@ -5196,7 +5231,7 @@ def filter_by(self: SelfSelect, **kwargs: Any) -> SelfSelect:
         from_entity = self._filter_by_zero()
 
         clauses = [
```
Comment (sqla-tester) [inline lib/sqlalchemy/sql/selectable.py:5233]: **mike bayer** (zzzeek) wrote:

remove

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4325

File: lib/sqlalchemy/sql/selectable.py [hunk 24]
```diff
@@ -5919,8 +5958,8 @@ def with_only_columns(
         _MemoizedSelectEntities._generate_for_statement(self)
 
         self._raw_columns = [
```
Comment (sqla-tester) [inline lib/sqlalchemy/sql/selectable.py:5960]: **mike bayer** (zzzeek) wrote:

remove both

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4325

File: lib/sqlalchemy/sql/selectable.py [hunk 25]
```diff
@@ -6474,22 +6513,22 @@ class ScalarSelect(
 
     element: SelectBase
 
-    def __init__(self, element):
+    def __init__(self, element: Select[Any]) -> None:
         self.element = element
         self.type = element._scalar_type()
```
Comment (sqla-tester) [inline lib/sqlalchemy/sql/selectable.py:6519]: **mike bayer** (zzzeek) wrote:

remove

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4325

File: lib/sqlalchemy/sql/selectable.py [hunk 26]
```diff
@@ -6687,7 +6728,7 @@ def correlate(
         """
         e = self._clone()
         e.element = self._regroup(
```
Comment (sqla-tester) [inline lib/sqlalchemy/sql/selectable.py:6730]: **mike bayer** (zzzeek) wrote:

remoive

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4325

File: lib/sqlalchemy/sql/selectable.py [hunk 27]
```diff
@@ -6706,7 +6747,7 @@ def correlate_except(
 
         e = self._clone()
         e.element = self._regroup(
```
Comment (sqla-tester) [inline lib/sqlalchemy/sql/selectable.py:6749]: **mike bayer** (zzzeek) wrote:

remove

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4325

File: lib/sqlalchemy/sql/selectable.py [hunk 28]
```diff
@@ -6724,7 +6765,7 @@ def select_from(self: SelfExists, *froms: FromClause) -> SelfExists:
 
         """
         e = self._clone()
```
Comment (sqla-tester) [inline lib/sqlalchemy/sql/selectable.py:6767]: **mike bayer** (zzzeek) wrote:

remove

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4325

File: lib/sqlalchemy/sql/selectable.py [hunk 29]
```diff
@@ -6742,7 +6783,7 @@ def where(
...
```
Comment (sqla-tester) [inline lib/sqlalchemy/sql/selectable.py:6785]: **mike bayer** (zzzeek) wrote:

remove

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4325

File: lib/sqlalchemy/sql/selectable.py [hunk 30]
```diff
@@ -6846,9 +6892,11 @@ def bindparams(
         self.element = self.element.bindparams(*binds, **bind_as_values)
         return self
```
Comment (sqla-tester) [inline lib/sqlalchemy/sql/selectable.py:6894]: **mike bayer** (zzzeek) wrote:

remove

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4325

File: lib/sqlalchemy/sql/util.py [hunk 31]
```diff
@@ -1164,7 +1164,7 @@ def replace(
                         return None
                 return self.selectable  # type: ignore
             elif isinstance(col, Alias) and isinstance(
-                col.element, TableClause
+                col.element, TableClause  # type: ignore[attr-defined]
```
Comment (sqla-tester) [inline lib/sqlalchemy/sql/util.py:1167]: **mike bayer** (zzzeek) wrote:

remove

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4325

**Review by sqla-tester — COMMENTED** [review]
**mike bayer** (zzzeek) wrote:

OK  -

thanks for the efforts.   So, there's lots and lots of ignores here, and these are actually counterproductive as they are covering up areas where typing needs to be improved.  If we are going to enable strict typing for this module, we have to actually get it ...typed!  :)   adding ignores on hundreds of lines just hides the problems that need to be solved.

If you want to continue on this, remove *all* ignores that were added, and just push up what we have.   then we will go through and fix as many of those problems as possible. lots of these seem related to things like _construct() needing better typing, that is, something we can fix in one place that will enable everything else to work.

I can also gladly pick this up as some of the typing problems here might be tricky to solve, not sure yet.   

also maybe some issues will improve if you make sure to use at least python 3.10 with mypy/dymyp.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4325

**Comment by dzmitar** [comment]
All new "# type: ignore" were removed.

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision b8d5950c93fa85522a301c6eae695d9de01f7047 of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
Patchset b8d5950c93fa85522a301c6eae695d9de01f7047 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4325

**Comment by zzzeek** [comment]
thanks!  I will look into these sections to see if i can improve the errors.   however I am going away tomorrow through sunday so it might be next week.

File: lib/sqlalchemy/sql/selectable.py [hunk 32]
```diff
@@ -1549,7 +1564,7 @@ def _from_objects(self) -> List[FromClause]:
 
 
 class NoInit:
-    def __init__(self, *arg, **kw):
+    def __init__(self, *arg: Any, **kw: Any) -> NoReturn:
```
Comment (zzzeek) [inline lib/sqlalchemy/sql/selectable.py:1567]: this part, this is why all the _construct() calls are failing for mypy. 

Change this to:

```
def __init__(self, *arg: Any, **kw: Any):
```

that is, remove the "-> NoReturn" and leave it implicit.   Errors go from about 59 to 39 and are much simpler.  Can you make that change, then do a pass for some of the easier ones such as:

```
lib/sqlalchemy/sql/selectable.py:1654: error: Function is missing a return type annotation  [no-untyped-def]
lib/sqlalchemy/sql/selectable.py:1763: error: Function is missing a type annotation for one or more arguments  [no-untyped-def]
```

you can also add typing to the util functions that are outside:

```
lib/sqlalchemy/sql/selectable.py:4668: error: Call to untyped function "_expand_cloned" in typed context  [no-untyped-call]
lib/sqlalchemy/sql/selectable.py:4716: error: Call to untyped function "_cloned_intersection" in typed context  [no-untyped-call]
lib/sqlalchemy/sql/selectable.py:4717: error: Call to untyped function "_cloned_intersection" in typed context  [no-untyped-call]
lib/sqlalchemy/sql/selectable.py:4730: error: Call to untyped function "_cloned_difference" in typed context  [no-untyped-call]

```

**Review by zzzeek — CHANGES_REQUESTED** [review]
we can get much farther if we remove the " -> NoReturn" from `NoInit.__init__()`. mypy does not appear to be able to accommodate this.

File: lib/sqlalchemy/sql/selectable.py (hunk 32 above)
Comment (dzmitar) [reply lib/sqlalchemy/sql/selectable.py]: Thanks, I fixed NoReturn and lib/sqlalchemy/sql/selectable.py:1654,
lib/sqlalchemy/sql/selectable.py:1763
But I couldn't find a solution for the rest of mypy issues.
Perhaps we should accept the types we are sure about for now, and do the rest after the types have been added to other modules?

Comment (zzzeek) [reply lib/sqlalchemy/sql/selectable.py]: yes perhaps!  just want to pare down what we have to a few errors :)    that is I want to make sure each "ignore" is really something that's not worth fixing.

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision e2f8ddeac0b08feaad917285e988acf1e9465a26 of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
Patchset e2f8ddeac0b08feaad917285e988acf1e9465a26 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4325

**Comment by zzzeek** [comment]
I've updated the Gerrit only (not the PR) with a few more changes, and I have to make a set of changes in a separate gerrit first with which to make this dependent on.  the remaining changes require more editorial decisions to update the argument signature of some methods.    I can complete the changes from here out.

**Comment by sqla-tester** [comment]
**mike bayer** (zzzeek) wrote:

thanks very much for doing this!   this was hard and had lots of things requiring very deep knowledge of the internals.  great job!

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4325

**Comment by sqla-tester** [comment]
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4325 has been **merged**. Congratulations! :)

**Pull request closed without merging**